File: SWEA/D3/3282. 0／1 Knapsack/0／1 Knapsack.py
# 3282. 0/1 Knapsack <D3>

tc = int(input())

for t in range(1, tc + 1):
    # n : 물건 갯수 | k : 최대 부피
    n, k = map(int, input().split())
    dp = [[0] * (k + 1) for _ in range(n + 1)]
    items = [list(map(int, input().split())) for _ in range(n)]

    for i in range(1, n + 1):
        for j in range(1, k + 1):
            if items[i - 1][0] <= j:
                dp[i][j] = max(dp[i - 1][j], items[i - 1][1] + dp[i - 1][j - items[i - 1][0]])
            else:
                dp[i][j] = dp[i - 1][j]

    print(f'#{t} {dp[n][k]}')

File: SWEA/D3/3282. 0／1 Knapsack/0／1 Knapsack.java
// 03282. [diff_3] 0/1 Knapsack.
import java.io.*;
import java.util.StringTokenizer;

public class Solution {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;
    public static void main(String[] args) throws IOException {
        int TC = Integer.parseInt(br.readLine());
        for (int tc = 1; tc <= TC; tc++) {
            st = new StringTokenizer(br.readLine());
            int N = Integer.parseInt(st.nextToken());
            int K = Integer.parseInt(st.nextToken());
            int[][] arr = new int[N][2];
            int[][] dp = new int[N+1][K+1];
            for (int n = 1; n <= N; n++) {
                st = new StringTokenizer(br.readLine());
                int V = Integer.parseInt(st.nextToken());
                int C = Integer.parseInt(st.nextToken());
                for (int k = 1; k <= K; k++) {
                    if (V <= k) {
                        dp[n][k] = Math.max(dp[n-1][k-V] + C, dp[n-1][k]);
                    } else {
                        dp[n][k] = dp[n-1][k];
                    }

                }
            }
            bw.write("#" + tc + " " + dp[N][K] + "\n");
        }
        bw.flush();
        bw.close();
        br.close();
    }
}